python/python/bystro/covariance/_base_covariance.py

```diff
@@ -0,0 +1,549 @@
+        The observed data
+
+    idxs: np.array-like,(sum(p),)
+        The observation locations
+
+    Returns
+    -------
+    preds : np.array-like,(N_samples,p-\sum idxs)
+        The predicted values
+    """
+    covariance_sub = covariance[idxs == 1]
+    covariance_22 = covariance_sub[:, idxs == 1]
+    covariance_21 = covariance_sub[:, idxs == 0]
+
+    covariance_nonsub = covariance[idxs == 0]
+    covariance_11 = covariance_nonsub[:, idxs == 0]
+
+    beta_bar = la.solve(covariance_22, covariance_21)
+    Second_part = np.dot(covariance_21.T, beta_bar)
+
+    covariance_11 - Second_part
+
+    preds = np.dot(Xobs[:, idxs == 1], beta_bar)
+    return preds
+
+
+#########################################
+#########################################
+##                                     ##
+##  Gaussian Likelihood-based methods  ##
+##                                     ##
+#########################################
+#########################################
+
+
+def _conditional_score(covariance, X, idxs, weights=None):
+    """
+    Returns the predictive log-likelihood of a subset of data.
+
+    mean(log p(X[idx==1]|X[idx==0],covariance))
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    X : np.array-like,(N,sum(idxs))
+        The data
+
+    idxs: np.array-like,(sum(p),)
+        The observation locations
+
+    weights : np.array-like,(N,),default=None
+        The optional weights on the samples
+
+    Returns
+    -------
+    avg_score : float
+        Average log likelihood
+    """
+    if weights is None:
+        weights = np.ones(X.shape[0])
+    avg_score = np.mean(
+        weights * _conditional_score_samples(covariance, X, idxs)
+    )
+    return avg_score
+
+
+def _conditional_score_samples(covariance, X, idxs):
+    """
+    Return the conditional log likelihood of each sample, that is
+
+    log p(X[idx==1]|X[idx==0],covariance) = N(Sigma_10Sigma_00^{-1}x_0,
+                                    Sigma_11-Sigma_10Sigma_00^{-1}Sigma_01)
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    X : np.array-like,(N,p)
+        The data
+
+    idxs: np.array-like,(p,)
+        The observation locations
+
+    Returns
+    -------
+    scores : float
+        Log likelihood for each sample
+    """
+    covariance_sub = covariance[idxs == 1]
+    covariance_22 = covariance_sub[:, idxs == 1]
+    covariance_21 = covariance_sub[:, idxs == 0]
+
+    covariance_nonsub = covariance[idxs == 0]
+    covariance_11 = covariance_nonsub[:, idxs == 0]
+
+    beta_bar = la.solve(covariance_22, covariance_21)
+    Second_part = np.dot(covariance_21.T, beta_bar)
+
+    covariance_bar = covariance_11 - Second_part
+
+    mu_ = np.dot(X[:, idxs == 1], beta_bar)
+    scores = _score_samples(covariance_bar, X[:, idxs == 0] - mu_)
+
+    return scores
+
+
+def _conditional_parameters(covariance, idxs):
```

Comment:
unless a method is a @property, it's generally preferred to begin them with a verb, e.g. `_get_conditional_parameters` or similar

Reply:
Done.